### PR TITLE
ios: picture in picture implementation

### PR DESCRIPTION
**Pull request by davidliu**
First draft of PIP on iOS. Suggestions on what the API should look like are appreciated.

Related issue: #1226

TODO:

- [x] `stopAutomatically` implementation
- [x] `preferredSize` implementation
- [x] Handle overflows properly
- [x] Handle rotation
- [x] Don't render while not active.
- [x] Use CVPixelBufferPool
- [x] Use [sampleBufferRenderer](https://developer.apple.com/documentation/avfoundation/avsamplebufferdisplaylayer/4172599-samplebufferrenderer?language=objc) on iOS 17+
- [x] Manual picture in picture activation?
- [x] Fallback view implementation

### CODE REVIEW

File: ios/RCTWebRTC/I420Converter.m [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+- (void)unprepareForAccelerateConversion
+{
+    if (_conversionInfo != NULL) {
+        free(_conversionInfo);
+        _conversionInfo = NULL;
+    }
+}
+
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:52]:
```suggestion

- (void)createPixelBufferPoolWithWidth:(size_t)width height:(size_t)height {
    if (_pixelBufferPool != NULL) {
        CVPixelBufferPoolRelease(_pixelBufferPool);
    }
    
    NSDictionary *pixelBufferAttributes = @{
        (id)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA),
        (id)kCVPixelBufferWidthKey: @(width),
        (id)kCVPixelBufferHeightKey: @(height),
        (id)kCVPixelBufferIOSurfacePropertiesKey: @{}
    };
    
    CVPixelBufferPoolCreate(kCFAllocatorDefault, NULL, (__bridge CFDictionaryRef)pixelBufferAttributes, &_pixelBufferPool);
}

```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 2]
```diff
@@ -0,0 +1,125 @@
+#import "I420Converter.h"
+#import "SampleBufferVideoCallView.h"
+#import <WebRTC/WebRTC.h>
+#import <Accelerate/Accelerate.h>
+
+
+@interface SampleBufferVideoCallView ()
+
+@property (nonatomic, retain) I420Converter *i420Converter;
+
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:10]:
```suggestion
@property (nonatomic, assign) CVPixelBufferPoolRef pixelBufferPool;
```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+@implementation SampleBufferVideoCallView
+
++ (Class)layerClass {
+    return [AVSampleBufferDisplayLayer class];
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    if (self = [super initWithFrame:frame]) {
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(layerFailedToDecode:) name:AVSampleBufferDisplayLayerFailedToDecodeNotification object:self.sampleBufferLayer];
+    }
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:22]:
```suggestion
        [self createPixelBufferPool];
    }
```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+- (AVSampleBufferDisplayLayer *)sampleBufferLayer {
+    return (AVSampleBufferDisplayLayer *)self.layer;
+}
+
+/** The size of the video frame. */
+- (void)setSize : (CGSize)size{
+    
+}
+
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:40]:
```suggestion

- (void)createPixelBufferPool {
    NSDictionary *pixelBufferAttributes = @{
        (id)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA),
        (id)kCVPixelBufferWidthKey: @(1000),
        (id)kCVPixelBufferHeightKey: @(1000),
        (id)kCVPixelBufferIOSurfacePropertiesKey: @{}
    };
    
    CVPixelBufferPoolCreate(kCFAllocatorDefault, NULL, (__bridge CFDictionaryRef)pixelBufferAttributes, &_pixelBufferPool);
}

```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    // TODO: handle rotations
+    // Create CMSampleBuffer
+    CMSampleBufferRef sampleBuffer;
+    CMSampleBufferCreateForImageBuffer(kCFAllocatorDefault, pixelBuffer, true, nil, nil, formatDescription, &timingInfo, &sampleBuffer);
+    
+    CFArrayRef attachments = CMSampleBufferGetSampleAttachmentsArray(sampleBuffer, YES);
+    CFMutableDictionaryRef dict = (CFMutableDictionaryRef)CFArrayGetValueAtIndex(attachments, 0);
+
+    CFDictionarySetValue(dict, kCMSampleAttachmentKey_DisplayImmediately, kCFBooleanTrue);
+
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:90]:
```suggestion
    
    CVPixelBufferRelease(pixelBuffer); // Release the pixel buffer back to the pool
    
```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        if(err != kvImageNoError) {
+            NSLog(@"Error when preparing i420Converter: %ld", err);
+            return NULL;
+        }
+        
+        _i420Converter = converter;
+    }
+    
+    return [_i420Converter convertI420ToPixelBuffer:i420Buffer];
+}
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:118]:
```suggestion
    if (_pixelBufferPool == NULL) {
        [self createPixelBufferPool];
    }
    
    CVPixelBufferRef pixelBuffer;
    CVReturn status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, _pixelBufferPool, &pixelBuffer);
    
    if (status != kCVReturnSuccess) {
        NSLog(@"Error creating pixel buffer from pool: %d", status);
        return NULL;
    }
    
    CVPixelBufferRef convertedPixelBuffer = [_i420Converter convertI420ToPixelBuffer:i420Buffer];
    CVPixelBufferRelease(pixelBuffer); // Release the pixel buffer back to the pool
    
    return convertedPixelBuffer;
}
```

File: ios/RCTWebRTC/SampleBufferVideoCallView.m [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+-(void)dealloc {
+    [_i420Converter unprepareForAccelerateConversion];
+}
```
Comment (momenthana) [inline ios/RCTWebRTC/SampleBufferVideoCallView.m:122]:
```suggestion
    if (_pixelBufferPool) {
        CVPixelBufferPoolRelease(_pixelBufferPool);
    }
}
```

File: ios/RCTWebRTC/I420Converter.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#import <UIKit/UIKit.h>
+#import <Accelerate/Accelerate.h>
+#import <WebRTC/WebRTC.h>
+
+@interface I420Converter : NSObject
+
+- (vImage_Error)prepareForAccelerateConversion;
+- (void)unprepareForAccelerateConversion;
+- (CVPixelBufferRef)convertI420ToPixelBuffer:(RTCI420Buffer *)buffer;
+
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.h:19]:
```suggestion
- (void)createPixelBufferPoolWithWidth:(size_t)width height:(size_t)height;

```

File: ios/RCTWebRTC/I420Converter.m [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+//  Created by Boisy Pitre on 5/21/16.
+//  Copyright © 2016 Twilio. All rights reserved.
+//
+
+#import "I420Converter.h"
+
+@interface I420Converter()
+
+@property (nonatomic, assign) vImage_YpCbCrToARGB *conversionInfo;
+
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:15]:
```suggestion
@property (nonatomic, assign) CVPixelBufferPoolRef pixelBufferPool;
```

File: ios/RCTWebRTC/I420Converter.m [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    return error;
+}
+
+- (void)unprepareForAccelerateConversion
+{
+    if (_conversionInfo != NULL) {
+        free(_conversionInfo);
+        _conversionInfo = NULL;
+    }
+}
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:51]:
```suggestion
    if (_pixelBufferPool != NULL) {
        CVPixelBufferPoolRelease(_pixelBufferPool);
        _pixelBufferPool = NULL;
    }
}
```

File: ios/RCTWebRTC/I420Converter.m [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+- (CVPixelBufferRef)convertI420ToPixelBuffer:(RTCI420Buffer *)buffer
+{
+    if (_conversionInfo == nil) {
+        NSLog(@"%@: not prepared", NSStringFromSelector(_cmd));
+        return NULL;
+    }
+
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:59]:
```suggestion
    
    if (_pixelBufferPool == NULL) {
        [self createPixelBufferPoolWithWidth:buffer.width height:buffer.height];
    }
    
```

File: ios/RCTWebRTC/I420Converter.m [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+                                           (id)kCVPixelBufferIOSurfacePropertiesKey,
+                                           nil
+    ];
+
+    CVReturn status = CVPixelBufferCreate(kCFAllocatorDefault,
+                                          buffer.width,
+                                          buffer.height,
+                                          kCVPixelFormatType_32BGRA,
+                                          (__bridge CFDictionaryRef)pixelBufferAttributes,
+                                          &pixelBuffer);
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:72]:
```suggestion
    CVReturn status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, _pixelBufferPool, &pixelBuffer);
```

File: ios/RCTWebRTC/I420Converter.m [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    // Do the conversion.
+    
+    uint8_t permuteMap[4] = {3, 2, 1, 0}; // BGRA
+    vImage_Error convertError = vImageConvert_420Yp8_Cb8_Cr8ToARGB8888(&yPlaneBuffer, &uPlaneBuffer, &vPlaneBuffer, &destinationImageBuffer, self.conversionInfo, permuteMap, 255, kvImageNoFlags);
+    
+    CVPixelBufferUnlockBaseAddress(pixelBufferRef, 0);
+    
+    return convertError;
+}
+
```
Comment (momenthana) [inline ios/RCTWebRTC/I420Converter.m:129]:
```suggestion

- (void)dealloc {
    [self unprepareForAccelerateConversion];
}

```

**Review by momenthana — COMMENTED** [review]
It's nice to have this part worked on!!! @davidliu 
I've tested this to help out a little bit, and I've written a part that works with CVPixelBufferPool.
It's a memory leak problem I faced a long time ago, and this is how I solve it.
https://github.com/momenthana/client-sdk-react-native/commit/56309f6b39124e6a875de265dcbc38a4d5ed8a55

before
before
Before and after
after
I solved it by applying Use CVPixelBufferPool to PR like task list. tested it with a physical iPad.

**Comment by davidliu** [comment]
Ready for review. I haven't run validation on remote tracks with this implementation yet, but otherwise should be good. Lemme know if there's any other basic functionality that should be added.

[Gist for GUMTestApp's App.js with example usage.](https://gist.github.com/davidliu/53727633ce522f3b67b900927df52744)

**Comment by momenthana** [comment]
@davidliu Audio session may complicate the processing of events for the stop button in picture, but the idea is to add an area for processing the PIP so that the image stops and streams no longer when the stop button is pressed.
How about creating AVPlayer display mode for live video like YouTube app for users who make streaming apps rather than conferences?
I suggest give some settings about displaying a button that doesn't go straight back to foreground when press PIP.

> The "실시간" displayed on the Korean screen is displayed as "Live" in English.

IMG_2783

I don't know if it's only possible to implement it using MPNowPlayingSession. Just a simple opinion.
The MPNowPlayingSession-related library I experienced was `react-native-music-control`. I don't know if controlling this is directly related to PIP.

```ts
      // stop and play control experience using react-native-music-control

      MusicControl.on(Command.play, () => {
        console.log("playRemoteControl");
        videoRemoteTrackPublication?.setEnabled(true);
        audioRemoteTrackPublication?.setEnabled(true);
      });
      MusicControl.on(Command.pause, () => {
        console.log("pauseRemoteControl");
        videoRemoteTrackPublication?.setEnabled(false);
        audioRemoteTrackPublication?.setEnabled(false);
      });
      MusicControl.on(Command.stop, () => {
        console.log("stopRemoteControl");
      });

      await AudioSession.startAudioSession();
      AudioSession.setAppleAudioConfiguration({
        audioCategory: "playback",
        audioCategoryOptions: ["interruptSpokenAudioAndMixWithOthers"],
        audioMode: "moviePlayback",
      });
      await AudioSession.showAudioRoutePicker();

      MusicControl.setNowPlaying({
        title: "Title",
        artwork: "",
        artist: "Hana",
        album: "Pinto",
        description: "",
        color: 0xffffff,
        colorized: true,
        notificationIcon: "my_custom_icon",
        isLiveStream: true,
      });
      MusicControl.enableControl("play", true);
      MusicControl.enableControl("pause", true);
```

image

**Comment by davidliu** [comment]
@momenthana unless someone comes up with a way to shunt the video frames into an AVPlayer, that's a no-go. It's beyond my ability to do. I'm not an iOS dev, so I'll need some example code to reference against.

**Comment by saghul** [comment]
Sorry for the delay @davidliu . I plan on taking a look next week!

**Comment by saghul** [comment]
> Ready for review. I haven't run validation on remote tracks with this implementation yet, but otherwise should be good. Lemme know if there's any other basic functionality that should be added.
> 
> [Gist for GUMTestApp's App.js with example usage.](https://gist.github.com/davidliu/53727633ce522f3b67b900927df52744)

Could you please incorporate this in the PR, perhaps wth a button to go into PiP mode? 🙏

**Comment by saghul** [comment]
@davidliu Quick update. @Calinteodor and I are testing this by integrating it into Jitsi Meet to see fi the API works there.

**Comment by saghul** [comment]
There is nothing special to do on Android, put your activity in pip mode and adapt to the size change in your app.

**Comment by Spharian** [comment]
> There is nothing special to do on Android, put your activity in pip mode and adapt to the size change in your app.

That's true, but some applications implement auto PiP mode when going to homescreen on Android for example, that's more intuitive and faster than putting activity in PiP mode. Do you think it's possible to implement?

**Comment by Spharian** [comment]
> You can enter PiP mode programatically in the "on user leave hint" activity lifecycle method.

Ooh. Do you think I should use something like https://github.com/adkaushik/react-native-pip-android ?
Thanks for your precious time and answers.

**Comment by Spharian** [comment]
For those wondering, used https://github.com/EdgarJMesquita/expo-pip and it worked like a charm for Android. Only waiting for this merge for iOS and our users will be very happy haha.

Thanks @saghul

File: ios/RCTWebRTC/PIPController.m [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+- (BOOL)startAutomatically {
+    return _pipController.canStartPictureInPictureAutomaticallyFromInline;
+}
+
+- (void)setStartAutomatically:(BOOL)value {
+    _pipController.canStartPictureInPictureAutomaticallyFromInline = value;
+}
+
+- (void)togglePIP {
```
[the remark this answers is not in the document]
Comment (davidliu) [reply ios/RCTWebRTC/PIPController.m:138]: Ughhhh, forgot to remove the `touchesEnded` trigger I was using for debugging purposes. It's in the RTCVideoViewManager.m (even appropriately labelled with `do not commit`).

[the remark this answers is not in the document]
Comment (Calinteodor) [reply ios/RCTWebRTC/PIPController.m:138]: Found it. Can you remove it, please? When you can, no rush.

**Comment by arekkubaczkowski** [comment]
great job guys! looking forward for releasing it!

**Comment by saghul** [comment]
@davidliu We got sidetracked so haven't had the time to fuklly integrate it in Jitsi Meet, but I trust we can make changes if necessary at a later stage and I didn't want to hold this any longer.

Excellent work mate!

**Comment by ponikar** [comment]
Any updates on release?

**Pull request merged**